packages/utilities/src/asAsync.tsx

```diff
@@ -0,0 +1,103 @@
+/**
+ * asAsync - a HOC for async loading components.
+ *
+ * Usage:
+ *
+ * const AsyncDialog = asAsync({
+ *   load: () => import('Dialog'),
+ *   exportName: 'Dialog', // defaults to 'default'
+ *   placeholder: () => <div>I am a placeholder</div>
+ * });
+ *
+ * React.render(domElement, <AsyncDialog { ...dialogProps } />);
+ */
+
+import * as React from 'react';
+
+interface IAsAsyncOptions<TProps> {
+  /**
+   * Callback which returns a promise resolving an object which exports the component.
+   */
+  load: () => Promise<{ [key: string]: React.ReactType<TProps> }>;
+
+  /**
+   * The exportName to resolve from the result, defaults to "default".
+   */
+  exportName?: string;
+
+  /**
+   * An optional JSX rendering for placeholder content.
+   */
+  placeholder?: React.ReactType;
```

Comment:
This is awkward because the consumer can't alter the placeholder contents at render time. That's the advantage of having it be exposed as a prop on the exposed component.

Reply:
That's true, but it also complicates the code and creates some opportunities for collisions... e.g. if I promote this to props, then what if I make, say, SearchBox which also has a "placeholder" prop? What is truth in this case? then it forces me to move the props into a props prop, which makes conversion messier. Or I go to a render component, which is also a little messy to consume.

Reply:
@ThomasMichon you ok with this concession? I would like to keep the bundle size minimal and feature set pretty clear. I removed the exportName to keep the type safety but would rather keep the placeholder config like it is. If you have a concern here, or a real scenario you need to support, let's evaluate an option. e.g. It's certainly possible to create a second thing to abstract dynamic loading scenarios, which just renders placeholder when render returns null.
```
<div>
   <DynamicLoadThing render={<AsyncThing ... /> } placeholder={ ... } />
</div>
```